### PR TITLE
Add quoted scalar and multi line Style configurable support

### PR DESCRIPTION
**Pull request by nefilim**
https://github.com/charleskorn/kaml/issues/245

### CODE REVIEW

File: src/jvmMain/kotlin/com/charleskorn/kaml/YamlOutput.kt
Comment (charleskorn) [inline src/jvmMain/kotlin/com/charleskorn/kaml/YamlOutput.kt:90]: Is the default value needed?

```suggestion
    private fun emitQuotedScalar(value: String, scalarStyle: ScalarStyle) = emitScalar(value, scalarStyle)
```

**Review by charleskorn — CHANGES_REQUESTED** [review]
Changes so far look good to me. Could you please add some tests for this? Once they're in place I'll be happy to merge this.

File: src/commonMain/kotlin/com/charleskorn/kaml/YamlConfiguration.kt
Comment (charleskorn) [inline src/commonMain/kotlin/com/charleskorn/kaml/YamlConfiguration.kt:44]: Should this default to matching `quotedScalarStyle`? So if the user sets `quotedScalarStyle`, then `mutliLineStringStyle` uses that value unless there's an explicit value given for `multiLineStringStyle`.

```suggestion
    internal val multiLineStringStyle: MultiLineStringStyle = quotedScalarStyle, // This won't compile due to the different types, but hopefully this makes the idea clear.
```

File: src/jvmMain/kotlin/com/charleskorn/kaml/YamlOutput.kt
Comment (nefilim) [reply src/jvmMain/kotlin/com/charleskorn/kaml/YamlOutput.kt:90]: it's used below for the emitting the type tag, I can remove the default value here and move it down there, or should that honour the `quotedScalarStyle` also?

Comment (charleskorn) [reply src/jvmMain/kotlin/com/charleskorn/kaml/YamlOutput.kt:90]: Ah, I'd missed that usage. I'd remove the default value and make it explicit in the caller.

**Comment by nefilim** [comment]
haven't noticed any existing test for `YamlOutput` .. am I missing them? Should I add to `JvmYamlWritingTest`?

**Comment by charleskorn** [comment]
> haven't noticed any existing test for `YamlOutput` .. am I missing them? Should I add to `JvmYamlWritingTest`?

`YamlWritingTest` would be the best place for these tests I think.

File: src/commonMain/kotlin/com/charleskorn/kaml/YamlConfiguration.kt
Comment (charleskorn) [inline src/commonMain/kotlin/com/charleskorn/kaml/YamlConfiguration.kt:44]: Should we call this `singleLineStringStyle` instead? That might make the distinction between this and `multiLineStringStyle` clearer.

File: src/commonMain/kotlin/com/charleskorn/kaml/YamlConfiguration.kt
Comment (charleskorn) [inline src/commonMain/kotlin/com/charleskorn/kaml/YamlConfiguration.kt:88]: Should we just have one type to cover both of these? There's no harm in allowing someone to emit single line strings with the literal style.

File: src/commonMain/kotlin/com/charleskorn/kaml/YamlConfiguration.kt
Comment (nefilim) [reply src/commonMain/kotlin/com/charleskorn/kaml/YamlConfiguration.kt:44]: this is what I had originally but it's used for characters and enums also, your call

File: src/commonMain/kotlin/com/charleskorn/kaml/YamlConfiguration.kt
Comment (nefilim) [reply src/commonMain/kotlin/com/charleskorn/kaml/YamlConfiguration.kt:88]: I would REALLY prefer this not to be the case, it really reduces the readability of the produced YAML and it almost doubles in length of the document when every single strings is in block literal format

File: src/commonMain/kotlin/com/charleskorn/kaml/YamlConfiguration.kt
Comment (charleskorn) [reply src/commonMain/kotlin/com/charleskorn/kaml/YamlConfiguration.kt:44]: Let's go with `singleLineStringStyle`.

File: src/jvmTest/kotlin/com/charleskorn/kaml/JvmYamlWritingTest.kt
Comment (charleskorn) [inline src/jvmTest/kotlin/com/charleskorn/kaml/JvmYamlWritingTest.kt:68]:
```suggestion
                        """
                            |"hello\nworld\nhow are | you?\n"
                            |
                        """.trimMargin()
```

Comment (charleskorn) [reply src/jvmTest/kotlin/com/charleskorn/kaml/JvmYamlWritingTest.kt:68]: This might make the intention a bit clearer (both here and in the other tests)

**Comment by charleskorn** [comment]
> I've added some tests, not sure why a new line is being added to every string?

Not 100% sure on this, but it's existing behaviour, so let's not add more to this PR. If it's concerning you, feel free to open another issue or raise a PR to fix it.

> ps. highly recommend you check out https://kotest.io/

I'm a huge fan of Kotest, just haven't had time to replace Spek with Kotest in here. PRs are always welcome :)

**Comment by peterfigure** [comment]
there's one spotless formatting violation left in the tests, not familiar with this plugin, can you ignore it perhaps?

**Comment by charleskorn** [comment]
> there's one spotless formatting violation left in the tests, not familiar with this plugin, can you ignore it perhaps?

Are you referring to this issue?

```
> The following files had format violations:
      src/jvmTest/kotlin/com/charleskorn/kaml/JvmYamlWritingTest.kt
          @@ -52,7 +52,7 @@
           ·····························hello
           ·····························world
           ·····························how·are·|·you?
          -························
          +
           ························""".trimIndent()
           ····················)
           ················}
  Run './gradlew :spotlessApply' to fix these violations.
```

If so, running `./gradlew spotlessApply` should fix the issue. It's complaining about the fact that that line is just whitespace characters.

**Comment by charleskorn** [comment]
Thanks @nefilim and @peterfigure!

**Pull request merged**
